legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/PureLSPGrammarExtension.java

```diff
@@ -216,9 +225,42 @@ public static FunctionLegendExecutionResult newResult(String id, Type type, Stri
     }
 
     @Override
-    public LegendExecutionResult executeLegendTDSRequest(SectionState sectionState, FunctionTDSRequest request)
+    public LegendExecutionResult executeLegendTDSRequest(SectionState section, FunctionTDSRequest request)
     {
-        return (executeFunction(sectionState, request.getEntity(), request.getInputParameters())).iterator().next();
+        CompileResult compileResult = getCompileResult(section);
+        String entityPath = request.getEntity();
+        if (compileResult.hasException())
+        {
+            return errorResult(compileResult.getException(), entityPath);
+        }
+        MutableList<LegendExecutionResult> results = Lists.mutable.empty();
+        PureModel pureModel = compileResult.getPureModel();
+        Function function = (Function) compileResult.getPureModelContextData().getElements().stream().filter(e -> e.getPath().equals(entityPath)).collect(Collectors.toList()).get(0);
+        ValueSpecification funcBody = function.body.get(0);
+        try
+        {
+            TDSGroupBy groupBy = request.getRequest().getGroupBy();
+            for (int index = 0; index < groupBy.getGroupKeys().size(); index++)
+            {
+                Filter groupFilter = new Filter(groupBy.getColumns().get(index), ColumnType.String, FilterOperation.EQUALS, groupBy.getGroupKeys().get(index));
+                request.getRequest().getFilter().add(groupFilter);
+            }
+            LegendTDSRequestLambdaBuilder.processFilterOperations(function, request.getRequest().getFilter());
+            LegendTDSRequestLambdaBuilder.processGroupByOperations(function, request.getRequest().getGroupBy(), request.getRequest().getColumns());
+            LegendTDSRequestLambdaBuilder.processSortOperations(function, request.getRequest().getSort());
```

Comment:
The code change the function that already exists on the PMCD.  This will have side effects, as now the original function does not reflect the proper state, and the concurrent behavior we have today, another thread could see the incorrect Function state.

We would need to copy/clone the function, or even better make these methods return a new function all the time, so the input is treated as immutable.